ENH: add TPR position and velocity read support

tylerjereddy opened the pull request:
Fixes gh-464.

* Add support for reading positions and velocities from GMX `.tpr` files over the full range of `tpx` versions supported by regular `.tpr` topology parsing (`58` through `137`).  We've had an issue open to add this capability for almost a decade now.

Notes for reviewers:

1. Suggestions for removing the use of `global` to track TPR file precision are appreciated.
2. Performance/duplication considerations -- as Richard noted in the cognate issue, the way our topology and coordinate handling code is organized, it may make sense to keep the separation here, but there should at least be opportunity to reduce code duplication if not completely avoid re-reading the topology to seek to the coordinate positions); that said, our `.tpr` reading performance is atrocious anyway and my attempts to fix at gh-4098 have stalled for more than a year, so I suggest we defer performance (and probably even duplication) considerations until after the capability and tests are cemented in.
3. Please help me identify TPR testing scenarios that are missing, and ideally suggest the TPR file(s) I should use to test them.
4. I'll probably deal with `CHANGELOG` and docs changes when it looks like we're close to ready to merge, but I still appreciate suggestions on i.e., locations where doc changes will be needed.




PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

## Developers certificate of origin
- [X] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--4873.org.readthedocs.build/en/4873/

<!-- readthedocs-preview mdanalysis end -->

tylerjereddy committed:
ENH: add TPR position read support

* Add support for reading positions from
GMX `.tpr` files at `tpx` version `134`.

pep8speaks commented:
Hello @tylerjereddy! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`package/MDAnalysis/coordinates/TPR.py`](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/package/MDAnalysis/coordinates/TPR.py):

> [Line 59:80](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/package/MDAnalysis/coordinates/TPR.py#L59): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 61:80](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/package/MDAnalysis/coordinates/TPR.py#L61): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)
> [Line 75:80](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/package/MDAnalysis/coordinates/TPR.py#L75): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)





* In the file [`testsuite/MDAnalysisTests/coordinates/test_tpr.py`](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py):

> [Line 15:80](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L15): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (119 > 79 characters)
> [Line 16:23](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L16): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 22:5](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L22): [E124](https://duckduckgo.com/?q=pep8%20E124) closing bracket does not match visual indentation
> [Line 24:23](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L24): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 30:5](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L30): [E124](https://duckduckgo.com/?q=pep8%20E124) closing bracket does not match visual indentation
> [Line 32:16](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L32): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 33:19](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L33): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 33:33](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L33): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 34:20](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L34): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 34:34](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L34): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 38:5](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L38): [E124](https://duckduckgo.com/?q=pep8%20E124) closing bracket does not match visual indentation
> [Line 40:21](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L40): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 41:19](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L41): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 41:33](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L41): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 42:19](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L42): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 42:33](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L42): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 44:20](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L44): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 46:5](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L46): [E124](https://duckduckgo.com/?q=pep8%20E124) closing bracket does not match visual indentation
> [Line 47:14](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L47): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 48:19](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L48): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 48:33](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L48): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 49:20](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L49): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 49:34](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L49): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 53:5](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L53): [E124](https://duckduckgo.com/?q=pep8%20E124) closing bracket does not match visual indentation
> [Line 54:14](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L54): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 55:19](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L55): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 55:33](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L55): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 56:20](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L56): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 56:34](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L56): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 60:5](https://github.com/MDAnalysis/mdanalysis/blob/7805923fcf5df2458b4fbc90dc6036b9e2e0ef65/testsuite/MDAnalysisTests/coordinates/test_tpr.py#L60): [E124](https://duckduckgo.com/?q=pep8%20E124) closing bracket does not match visual indentation





##### Comment last updated at 2024-12-31 17:39:01 UTC

tylerjereddy commented on package/MDAnalysis/coordinates/TPR.py:
```diff
@@ -0,0 +1,73 @@
+# -*- Mode: python; tab-width: 4; indent-tabs-mode:nil; coding:utf-8 -*-
+# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4
```
header needs expanding to full version

tylerjereddy commented on package/MDAnalysis/coordinates/TPR.py:
```diff
@@ -0,0 +1,73 @@
+    # or perhaps combine the topology and coordinate reading
+    # with some inheritance shenanigans?
+    format = "TPR"
+    units = {"length": "nm"}
```
this can be expanded when velocity support is added later

tylerjereddy commented on package/MDAnalysis/coordinates/TPR.py:
```diff
@@ -0,0 +1,73 @@
+
+        if th.bX:
+            self.ts._pos = np.asarray(tpr_utils.ndo_rvec(data, th.natoms),
+                                      dtype=np.float32)
```
I don't anticipate much trouble adding velocity information capture as well, should basically be a repeat of this for `bV`. I see no evidence of additional binary seek striding after `bX` capture in GMX C++ src.

tylerjereddy commented on package/MDAnalysis/topology/tpr/utils.py:
```diff
@@ -423,6 +423,48 @@ def do_mtop(data, fver, tpr_resid_from_one=False):
+        im_excl_grp_size = data.unpack_int()
+        ndo_int(data, im_excl_grp_size)
+        # TODO: why is this needed?
+        data.unpack_int()
```
The above can probably be cleaned up a bit.. probably has some unused vars, etc.

It was produced by careful `printf`-ing the GMX source as you might expect. Expanding to support other `tpx` versions may not be too bad, although this was fairly time consuming to draft.

tylerjereddy commented on package/MDAnalysis/topology/tpr/utils.py:
```diff
@@ -423,6 +423,48 @@ def do_mtop(data, fver, tpr_resid_from_one=False):
+        # api/legacy/include/gromacs/topology/topology_enums.h
+        # worst case scenario we hard code it based on
+        # tpx/GMX version?
+        SimulationAtomGroupType_size = 10
```
On latest GMX `main` branch, this is the data structure in question:

```C++
enum class SimulationAtomGroupType : int
{
    TemperatureCoupling,
    EnergyOutput,
    Acceleration,
    Freeze,
    User1,
    User2,
    MassCenterVelocityRemoval,
    CompressedPositionOutput,
    OrientationRestraintsFit,
    QuantumMechanics,
    Count
};
```

tylerjereddy commented on testsuite/MDAnalysisTests/coordinates/test_tpr.py:
```diff
@@ -0,0 +1,34 @@
+    u = mda.Universe(tpr_file)
+    assert_allclose(u.atoms.positions[0, ...], exp_first_atom)
+    assert_allclose(u.atoms.positions[-1, ...], exp_last_atom)
+    assert_equal(u.atoms.positions.shape, exp_shape)
```
I think we have specific conventions for coordinate reader testing, but for now this is where I've started. Should be easy to expand to include velocities as well.

Cases with only positions and no velocities (etc.) may also be sensible to add, on top of older `tpx` files...

orbeckst replied:
This looks already good and is quite clean and human-readable. Extra thanks for the comments that list the tpx versions.

orbeckst replied:
I had a quick look at our test suite and many SingleFrameReader tests are pretty much bespoke so I'd say what you wrote here is at least as good as what's already there.

If we wanted to use `base.BaseReaderTest` https://github.com/MDAnalysis/mdanalysis/blob/d412c9a9a56312c1bd4e33e6dd3afc4cec7783ca/testsuite/MDAnalysisTests/coordinates/base.py#L263 (e.g., like the TestGROReader https://github.com/MDAnalysis/mdanalysis/blob/d412c9a9a56312c1bd4e33e6dd3afc4cec7783ca/testsuite/MDAnalysisTests/coordinates/test_gro.py#L200 ) then we could just pick one of the test TPRs and create the appropriate BaseReference.

(The old  `base._SingleFrameReader` https://github.com/MDAnalysis/mdanalysis/blob/d412c9a9a56312c1bd4e33e6dd3afc4cec7783ca/testsuite/MDAnalysisTests/coordinates/base.py#L46 could directly use the AdK TPR file (`datafiles.TPR`) but I probably would stay away from these legacy tests.)

RMeli commented:
@tylerjereddy since this PR is composed of two new files, and the modifications are all in one place, would you mind if I go ahead with #4849? I don't think there would be significant conflicts.

tylerjereddy commented:
ok

RMeli commented:
Thanks @tylerjereddy. It is now merged.

tylerjereddy committed:
ENH: PR 4873 revisions

* Expand `TPRReader()` support to include velocity handling,
and add tests/functionality for an additional tpx version (`133`).

[ci skip] [skip azp]

tylerjereddy commented:
I drafted in support/testing for handling velocities as well, plus support/testing for reading positions/velocities from one more `tpx` version (`133`). I'm trying to skip the CI when I push in for the next little bit (looks like Azure still ran though..).

Need to poke around the `.tpr` test files to see if there are some more interesting ones as well re: non-zero velocities, etc.

tylerjereddy committed:
ENH, TST: PR 4873 revisions

* Add `.tpr` position/velocity reading support back
to GMX 2023 (tpx version 129), and associated testing.

* Add a `.tpr` position/velocity reading test
case that has non-zero velocities and many more
atoms. This test case currently passes for position
retrieval but not for velocity retrieval.

[ci skip] [skip azp] [azp skip] [skip ci]

tylerjereddy commented:
I was able to extend support and testing for `.tpr` position/velocity reads back as far as GMX 2023 (tpx version `129`); GMX 2022rc1 (tpx `127`) will fail testing because it needs more binary file striding version checks that I'll need to study in the C++ code (not too hard, but time consuming).

I also modernized an old TPR file we had with non-zero velocities (and a large number of atoms) for cobrotoxin via `gmx convert-tpr` and then added a test case for this. It passes for positions, but not for velocities yet, so that was a good intuition to check non-zero velocity/larger system case. Since that was converted to `tpx` `134` it will be a little harder to sort out the problem there, likely requiring side-by-side `printf-vs-print` for GMX vs. us on the binary striding.

tylerjereddy committed:
BUG: PR 4873 revisions

* Fixed an issue where `TPRReader` used `ts._velocities`
instead of `ts.velocities` to assign the velocity array

* Needed to improve the precision of the expected velocity
values in `test_basic_read_tpr` after the above fix.

* `TPR_xvf_2024_4` was missing from `__all__`, causing
a test failure

tylerjereddy commented:
> I also modernized an old TPR file we had with non-zero velocities (and a large number of atoms) for cobrotoxin via gmx convert-tpr and then added a test case for this

This test case for non-zero velocities should be passing now, as should other tests, apart from the linter, after some more bug fixes.

The outstanding items here would likely be:

1. going back to support a wider range of `tpx` versions
2. probably a few more `tpr` files with non-zero velocities across generations should be tested

tylerjereddy committed:
ENH, MAINT: PR 4873 revisions

* Add preliminary support for tpx version 137 (GROMACS 2025.0). This
required additional bit unpacking based on empirical observations
on the `TPR_NNPOT_2025_0` test file. A regression test for position
and velocity reads from that test file was added, with expected
values from `gmx dump -s ala_nnpot_gmx_2025_0.tpr`.

* Fix up formatting complaints from `black` that were causing CI
failures.

tylerjereddy commented:
Updates from work today:

- Added support for reading positions and velocities from `tpx` version 137 (GROMACS `2025.0`), along with a regression test confirming position and velocity reads from a `.tpr` file from that version of GMX (expected values from `gmx dump` as usual)
- finally fixed up the `black` linter complaints

tylerjereddy committed:
ENH: PR 4873 revisions

* Add support and testing for `tpx` version 127 (GROMACS 2022),
including two regression tests for that `.tpr` format. Supporting
this file version required careful analysis of relative binary
file offsets.

tylerjereddy commented:
Added support for `tpx` version 127 today (GROMACS `2022`)--two regression tests with different numbers of atoms were used in this case because I had to do manual binary offset analysis vs. newer versions of the format to get the positions properly extracted.

Next up from our support table should probably be `tpx` 122 (GROMACS `2021`), maybe tomorrow.

tylerjereddy committed:
ENH: PR 4873 revisions

* Add support for velocity and coordinate reading from `.tpr`
files that are in tpx generation `122` (GROMACS 2021). Includes
two regression tests.

tylerjereddy commented:
For today, I added support for `tpx` version `122` (GROMACS `2021`) + two regression tests. Luckily, this was easier than the last two versions I added, no byte skipping shims needed.

For tomorrow, I'll see if I can add support for `tpx` version `119` (GROMACS `2020`).

tylerjereddy committed:
ENH: PR 4873 revisions

* Add support for position and velocity data reading from `.tpr` files
that correspond to `tpx` version 119 (GROMACS `2020`). Two regression
tests for `.tpr` files at this version of GMX were added.

tylerjereddy commented:
Today, added support for `tpx` version 119 (GROMACS `2020`), which required minor byte striding adjustments to pass the usual two regression tests.

For tomorrow, I'll aim to extend backward to `tpx` 116 (GROMACS `2019`).

tylerjereddy committed:
ENH: PR 4873 revisions

* Add support for reading positions and velocities from
`.tpr` files at `tpx` version 116 (GROMACS `2019`). Likewise
for `tpx` version 112 (GROMACS `2018`).

* Add regression tests for the above improvements.

tylerjereddy commented:
Today, support was added for tpx 116 (GMX 2019) and tpx 112 (GMX 2018). The latter required some non-trivial byte-stride analysis on the binary files.

For tomorrow, I'll take aim at tpx 110 (GMX 2016) if I can.

tylerjereddy committed:
ENH: PR 4873 revisions

* Add support for position and velocity reading from `.tpr`
files that correspond to `tpx` version 110 (GROMACS `2016`).
Likewise for `tpx` version 103 (GROMACS `5.1`).

* Add regression tests for the above improvements.

tylerjereddy commented:
Today, added support for `tpx` 110 (GROMACS `2016`) and `tpx` 103 (GROMACS `5.1`).

I'll try to check `tpx` 100 (GMX `5.x` before `5.1`) tomorrow.

tylerjereddy committed:
ENH: PR 4873 revisions

* Add position and velocity reading support for `.tpr`
files at `tpx` version 100 (for GROMACS `5.x` before `5.1`),
including regression tests.

tylerjereddy commented:
Today, added support for `tpx` version 100 (GMX `5.x` before `5.1`). No byte striding shims were needed.

Next up is `tpx` version 83 (GMX 4.6.x)--I did a quick check and that will require more analysis/byte striding.

tylerjereddy committed:
ENH: PR 4873 revisions

* Add support for reading positions and velocities from `.tpr`
files at `tpx` version 83 (GMX `4.6.0`, `4.6.1`). Likewise
for `tpx` versions 73 (GMX `4.5.x`) and 58 (GMX `4.0.x`).
Also added some shims for improved double precision support.

* Add regression tests for the above improvements.

tylerjereddy commented on package/MDAnalysis/topology/tpr/utils.py:
```diff
@@ -74,6 +74,8 @@
     Impropers,
 )
 
+global_precision = 4
```
suggestions for tracking TPR file precision in a way that doesn't use a `global` are appreciated

tylerjereddy commented:
Ok, this weekend I added position and velocity reading support for all remaining `tpx` versions that MDAnalysis can read topology from under normal circumstances.

I think this is ready for review now--I've updated the original comment above with some notes for reviewers. The linter will currently fail because I'm using a `global` construct to track TPR precision.

Suggestions for fixing that and anything else you find are appreciated.

orbeckst commented on package/MDAnalysis/coordinates/TPR.py:
```diff
@@ -0,0 +1,80 @@
+
+
+class TPRReader(base.SingleFrameReaderBase):
+    # TODO: reduce duplication with `TPRparser`;
+    # we could share some state for the position
+    # in the binary file to avoid re-reading topology
+    # or perhaps combine the topology and coordinate reading
+    # with some inheritance shenanigans?
```
I also feel that it would be good to avoid reading twice.

However, there's something to be said for starting with a clean separation, establishing correct functionality, and then thinking about optimization.

orbeckst replied:
Optional.

orbeckst commented on package/MDAnalysis/coordinates/TPR.py:
```diff
@@ -0,0 +1,80 @@
+        data = tpr_utils.TPXUnpacker(tprf)
+        try:
+            th = tpr_utils.read_tpxheader(data)  # tpxheader
+        except (EOFError, ValueError):
```
Eventually a test that is triggered by wrong format would be good – I _assume_ we do this for the topology parser somewhere???

orbeckst reviewed (commented):
Only had time to have an initial look.

For review it might already be useful to also have the entry for the coordinate formats table in the doc to summarize capabilities. My understanding is that this PR adds capabilities for reading positions and velocities from TPR files when doing `u = Universe(TPR)`, similar to what we can do with a single PDB (positions only) or GRO.

tylerjereddy committed:
DOC, TST, BUG: PR 4873 revisions

* Add TPR position and velocity reading support to the
coordinate reading support table in our docs, as requested
by reviewer.

* Add a regression test for the proper raising of an error
when we use a proper TPR topology file, but an improper format
TPR coordinate file. This also revealed a bug in `TPR.py`
where the `logger` was not initialized, which is now fixed
and guarded by the same regression test. Note that this
new regression test does correctly distinguish the error
message for an improper coordinate TPR, which is distinct
from an improper topology TPR (I adjusted the error message
to mention coordinate parsing specifically).

* It may be worth noting that before dealing with the above
point, I hadn't thought much about the `u = mda.Universe(tpr_file, tpr_file)`
incantation vs. extracting topology and coordinates from the single
argument. That said, the separated-argument incantation does appear to
work just fine as well, so we could consider i.e., parametrizing our TPR
coordinate/velocity testing infrastructure to probe both incantations
if desirable.

* The use of a `global` variable for storing tpx precision data
has been removed in favor of adjuting `do_mtop()` to accept
this information as a new argument. This addresses one of my
own review comments (and a linter complaint about globals).

tylerjereddy commented:
@orbeckst I took at stab at addressing your and my review comments, after having to rebase to deal with gh-5061. Pasting from the commit message below the fold:

<details>
<summary> Summary of latest revisions to PR 4873 </summary>

```
    DOC, TST, BUG: PR 4873 revisions
    
    * Add TPR position and velocity reading support to the
    coordinate reading support table in our docs, as requested
    by reviewer.
    
    * Add a regression test for the proper raising of an error
    when we use a proper TPR topology file, but an improper format
    TPR coordinate file. This also revealed a bug in `TPR.py`
    where the `logger` was not initialized, which is now fixed
    and guarded by the same regression test. Note that this
    new regression test does correctly distinguish the error
    message for an improper coordinate TPR, which is distinct
    from an improper topology TPR (I adjusted the error message
    to mention coordinate parsing specifically).
    
    * It may be worth noting that before dealing with the above
    point, I hadn't thought much about the `u = mda.Universe(tpr_file, tpr_file)`
    incantation vs. extracting topology and coordinates from the single
    argument. That said, the separated-argument incantation does appear to
    work just fine as well, so we could consider i.e., parametrizing our TPR
    coordinate/velocity testing infrastructure to probe both incantations
    if desirable.
    
    * The use of a `global` variable for storing tpx precision data
    has been removed in favor of adjuting `do_mtop()` to accept
    this information as a new argument. This addresses one of my
    own review comments (and a linter complaint about globals).
```

</details>

One remaining action item that your comments got me thinking about is the incantations like `u = mda.Universe(tpr_file, tpr_file)`. This does seem to work just fine, as the detailed review comment below the fold indicates, but there is perhaps a question of wanting to test that incantation as well. I could parametrize the current test to probe both incantations I suppose.

Anything else that we are worried about? Possibly assessing performance impact, though I believe we may be in agreement that going for the new functionality first and then dealing with performance issues later may be "ok."

orbeckst commented:
Super-quick comment: Looks as if you addressed all concerns. I'd merge correct code even if there's still optimization to be done. However, is there any performance regression for the cases that TPRs had been used for in the past, i.e., do your changes incur an overhead even if x/v are not needed as in `Universe(TPR, XTC)`?

orbeckst commented:
Checking that `u = mda.Universe(tpr_file, tpr_file)` and `u = mda.Universe(tpr_file)` both work would be good.

orbeckst commented on package/MDAnalysis/coordinates/TPR.py:
```diff
@@ -0,0 +1,85 @@
+logger = logging.getLogger("MDAnalysis.coordinates.TPR")
+
+import numpy as np
+
```
Needs some docs to be included in the sphinx doc tree. 

Include a versionadded 2.10.0.

orbeckst commented on package/MDAnalysis/topology/tpr/utils.py:
```diff
@@ -423,6 +423,79 @@ def do_mtop(data, fver, tpr_resid_from_one=False):
+        # from GMX source code:
+        # api/legacy/include/gromacs/topology/topology_enums.h
+        # worst case scenario we hard code it based on
+        # tpx/GMX version?
```
perhaps explicitly mention this PR for more background? Your annotation of the code is really useful and in this way anyone digging into this code can also refer back to them, such as your analysis of data structures etc.

tylerjereddy committed:
TST: PR 4873 revisions

* Adjust the testing in above PR such that we also probe coordinate
and velocity reading from TPR files using the "double incantation:"
`u = mda.Universe(tpr_file, tpr_file)`, and not just the single
incantation.

* Refer back to the above PR in the code comments for clarity,
based on reviewer feedback.

* Add a regression test for the raising of an appropriate error
when attempting to read positions and velocities from the beta
series of GROMACS 2020. There is part of the effort to improve
test coverage in the above PR.

* Add a simple regression test for the case of reading the topology
of a given system from a different version of TPR file than the
version of TPR containing the positions and velocities.

* Remove a guard for the presence of topology data in TPR
files--I'm not aware of any cases where the topology information
could reasonably be absent (no way to test this at the moment).

* Add full canonical file header to new files added in above PR.

* Update `CHANGELOG` to reflect the new functionality added in above PR.

* Add some initial/crude documentation to `TPR.py`, along with the
requested `versionadded` directive.

orbeckst reviewed (changes requested):
Looks very good at first glance.

Please also update CHANGELOG (and minor things such as the full canonical file header with copright etc).

Coverage is lower than average so there are probably some cases that haven't been tested. If you think that it's not worthwhile, add an explicit [`#pragma: no cover`](https://coverage.readthedocs.io/en/latest/excluding.html) to be explicit about our holes in testing.

My inline comments are just suggestions.

tylerjereddy commented:
Another thing I didn't test, but that actually works, is different versions of TPR for topology vs. x/v reading: `u = mda.Universe(TPR2020, TPR2024_4)`. That's perhaps slightly mind bending, but I suppose there is no reason it shouldn't work. I'm not sure how far I'd want to go in officially supporting/testing that at this time, but just noticed it while addressing comments locally. Maybe I'll just add a test for a single case like that for now.

orbeckst commented:
Yes, add a one off test, maybe it breaks in interesting ways in the future and reveals a real issue. Might eventually be useful when we optimize towards reading only once instead of separately for topology and x/v.

tylerjereddy commented:
I pushed in a bunch more revisions with a detailed commit message. Probably some kind of linter/docstring issue will crop up now.

Let me see what I can do about checking the performance implications for prior usage a bit.

tylerjereddy commented:
I'm having trouble running the `asv` benchmarks locally on both Mac and Linux. The sample benchmarks I drafted are below the fold (looks like this file may have annoyances with carriage returns as well?). My sample incantation was: `asv continuous -E virtualenv -v -e -b "TPRBench" develop treddy_issue_464`

<details>

```diff
--- a/benchmarks/benchmarks/topology.py
+++ b/benchmarks/benchmarks/topology.py
@@ -4,7 +4,8 @@ from MDAnalysis.guesser import DefaultGuesser
 
 try:
     from MDAnalysis.exceptions import NoDataError
-    from MDAnalysisTests.datafiles import GRO
+    from MDAnalysisTests.datafiles import (GRO, TPR_xvf_2024_4,^M
+                                           TPR2023, TPR406)^M
 except:
     pass
 
@@ -45,3 +46,26 @@ class BondsBench(object):
     def time_bonds(self, num_bonds):
         """Benchmark for calculating bonds"""
         self.u.bonds
+^M
+^M
+class TPRBench:^M
+    """^M
+    Benchmarks for reading in TPR files, for which we^M
+    support both topology and coordinate/velocity parsing.^M
+    """^M
+    # benchmark over a variety of TPR versions and^M
+    # system sizes^M
+    params = (TPR_xvf_2024_4, # 19385 atoms^M
+              TPR2023, # 2263 atoms^M
+              TPR406, # 2263 atoms^M
+             )^M
+    param_names = ["tpr_file"]^M
+^M
+^M
+    def time_single_incantation_tpr(self, tpr_file):^M
+        MDAnalysis.Universe(tpr_file)^M
+^M
+^M
+    def time_double_incantation_tpr(self, tpr_file):^M
+        # only possible from MDA 2.10.0 onward^M
+        MDAnalysis.Universe(tpr_file, tpr_file)^M
```

</details>

I see `Failed to build the project and import the benchmark suite.` shortly after `Successfully built MDAnalysis` in a verbose log... don't have time to dig further on that at the moment. If other folks regularly run the benchmarks in a venv setup I'm open to suggestions... it looks like `asv.conf.json` may also need some modernization (some projects need to customize specific build/install cmds in there too).

orbeckst committed:
Merge branch 'develop' into treddy_issue_464

orbeckst reviewed (approved):
Thank you for addressing my remaining comments. 

Anything related to ASV is a separate issue.

Sorry it took me 2 weeks to get back to your PR.

Assuming CI is happy, I'd like to squash-merge — thank you for a great new feature!

orbeckst committed:
add TPR to coordinate docs

orbeckst committed:
add TPR to coordinates toctree

orbeckst commented:
Just after reviewing I looked at the docs and realized that the new TPR coordinate docs were not linked. I quickly added the files... hopefully all looks good.

orbeckst commented on package/MDAnalysis/coordinates/TPR.py:
```diff
@@ -0,0 +1,133 @@
+-------
+
+.. autoclass:: TPRReader
+   :members:
```
```suggestion
   :members:
   :inherited-members:
```
Consistent with other docs.

orbeckst commented on package/doc/sphinx/source/documentation_pages/coordinates/TPR.rst:
```diff
@@ -0,0 +1,3 @@
+.. automodule:: MDAnalysis.coordinates.TPR
+   :members:
+   :inherited-members:
```
Explicitly list classes in coordinates/TPR.py
```suggestion
```

orbeckst reviewed (approved):
LGTM

(just doc formatting fixups)

orbeckst committed:
Apply suggestions from code review

orbeckst commented:
Congratulations @tylerjereddy for closing a decade-old issue!

(Btw, I carefully preserved and edited your log messages in the squash-commit as they contain a plethora of useful information.)
